packages/api/src/controllers/generate.ts

```diff
@@ -39,58 +167,49 @@ function registerGenerateHandler(
         return;
       }
 
+      const startedAt = Date.now();
       const apiUrl = pathJoin2(aiGatewayUrl, path);
-
-      let payload: BodyInit;
-      if (isJSONReq) {
-        payload = JSON.stringify({
-          model_id: defaultModel,
-          ...req.body,
+      const [body, request] = createPayload(req, isJSONReq, defaultModel);
+      let gatewayRes: Response, response: Buffer, error: any;
+      try {
+        gatewayRes = await fetchWithTimeout(apiUrl, {
+          method: "POST",
+          body,
+          timeout: AI_GATEWAY_TIMEOUT,
+          headers: isJSONReq ? { "content-type": "application/json" } : {},
         });
-      } else {
-        const form = new FormData();
-        if (!("model_id" in req.body)) {
-          form.append("model_id", defaultModel);
-        }
-        for (const [key, value] of Object.entries(req.body)) {
-          form.append(key, value);
-        }
 
-        if (!Array.isArray(req.files)) {
-          return res.status(400).json({
-            errors: ["Expected an array of files"],
-          });
-        }
-        for (const file of req.files) {
-          form.append(file.fieldname, file.buffer, {
-            filename: file.originalname,
-            contentType: file.mimetype,
-            knownLength: file.size,
-          });
+        response = await gatewayRes.buffer();
+      } catch (err) {
+        error = err;
+      } finally {
+        logAiGenerateRequest(
+          req.user.id,
+          req.project?.id,
+          type,
+          startedAt,
+          request,
+          gatewayRes?.status,
+          response,
+          error,
+        );
+        if (error) {
+          throw error;
         }
-        payload = form;
       }
 
-      const response = await fetchWithTimeout(apiUrl, {
-        method: "POST",
-        body: payload,
-        timeout: AI_GATEWAY_TIMEOUT,
-        headers: isJSONReq ? { "content-type": "application/json" } : {},
-      });
-
-      const body = await response.json();
-      if (!response.ok) {
-        logger.error(
-          `Error from generate API ${path} status=${
-            response.status
-          } body=${JSON.stringify(body)}`,
-        );
-      }
-      if (response.status >= 500) {
-        return res.status(500).json({ errors: [`Failed to generate ${name}`] });
+      if (gatewayRes.status >= 500) {
+        // We hide internal server error details from the user.
+        return res.status(500).json({ errors: [`Failed to generate ${type}`] });
       }
 
-      res.status(response.status).json(body);
+      // forward content metadata headers
```

Comment:
Actually this was necessary because I started treating the response as a buffer instead of always a `json()`. This was required just to get to the guarantee we don't miss information (like response content) when the AI Gateway gives a bad response that is not a JSON (like a literal `500 Internal Server Error`, could be some proxy, idk)

So previously, this code would call sth like `res.json()` that would headers appropriately. Now that we proxy the response verbatim with `buffer()` and `send()`, we need to proxy the content headers too.